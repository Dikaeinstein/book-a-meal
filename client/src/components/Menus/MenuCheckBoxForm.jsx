import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ProgressBar } from 'react-materialize';
import Loading from '../util/Loading';
import Label from '../util/Label';
import ConnectedMealsCheckboxList from './MealsCheckboxList';
import Paginate from '../util/Paginate';
import '../../static/sass/materialize.scss';

class MealsCheckBoxForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: this.props.defaultMenuName,
      mealIds: this.props.defaultMenuMealIds,
      checked: this.props.defaultMenuMealIds
        .reduce((obj, key) => ({ ...obj, [`meal-${key}`]: true }), {}),
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleCheck = this.handleCheck.bind(this);
  }

  handleChange(event) {
    this.setState({
      name: event.target.value,
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    this.props.handleSubmit({
      name: this.state.name,
      mealIds: this.state.mealIds,
    });
  }

  handleCheck(event) {
    if (event.target.checked) {
      // Extract mealId from input name
      const mealId = event.target.name.slice(5);
      this.setState({
        mealIds: [...this.state.mealIds, parseInt(mealId, 10)],
        checked: {
          ...this.state.checked,
          [event.target.name]: true,
        },
      });
    } else {
      const mealId = event.target.name.slice(5);
      const filteredMealIds = this.state.mealIds.filter(OldMealId =>
        OldMealId !== parseInt(mealId, 10));

      this.setState({
        mealIds: filteredMealIds,
        checked: {
          ...this.state.checked,
          [event.target.name]: false,
        },
      });
    }
  }

  render() {
    return (
      <form
        className="form card"
        onSubmit={this.handleSubmit}
      >
        <h6
          className="text-center"
          style={{ margin: '.5rem 0 1rem 0' }}
        >
          Select meals to add to menu
        </h6>
        <div>
          <Label
            className="label label-block font-weight-bold"
            htmlFor="name"
          >
            Menu Name:
          </Label>
          <input
            name="name"
            placeholder="Enter menu name"
            value={this.state.name}
            onChange={this.handleChange}
          />
        </div>
        <div>
          <h6
            style={{ margin: '.5rem 0' }}
            className="font-weight-bold"
          >
            Meals
          </h6>
          <ConnectedMealsCheckboxList
            meals={this.props.meals}
            checkedMeals={this.state.checked}
            handleCheck={this.handleCheck}
          />
          <Paginate
            onPageChange={this.props.onPageChange}
            nextUrl={this.props.nextUrl}
            previousUrl={this.props.previousUrl}
            style={{ marginTop: '0' }}
          />
        </div>
        <button
          value="Set Menu"
          type="submit"
          className="btn btn-default"
          style={{ margin: '0 auto 1rem auto' }}
        >
          {this.props.action} Menu
        </button>
        {this.props.isSubmitting ?
          <Loading text="setting menu">
            <ProgressBar />
          </Loading>
          : null}
        {!this.props.isSubmitting && this.props.error ?
          <div className="error">{this.props.error}</div>
          : null
        }
      </form>
    );
  }
}

MealsCheckBoxForm.defaultProps = {
  defaultMenuName: 'Menu for Today',
  defaultMenuMealIds: [],
};

MealsCheckBoxForm.propTypes = {
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  defaultMenuName: PropTypes.string,
  defaultMenuMealIds: PropTypes.arrayOf(PropTypes.number),
  handleSubmit: PropTypes.func.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  /* eslint react/require-default-props: 0 */
  error: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.objectOf(PropTypes.string),
  ]),
  action: PropTypes.string.isRequired,
  nextUrl: PropTypes.string,
  previousUrl: PropTypes.string,
  onPageChange: PropTypes.func.isRequired,
};

export default MealsCheckBoxForm;
