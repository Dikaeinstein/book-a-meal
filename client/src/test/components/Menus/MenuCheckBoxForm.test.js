import React from 'react';
import { shallow } from 'enzyme';
import MenuCheckBoxForm
  from '../../../components/Menus/MenuCheckBoxForm';

describe('<MenuCheckBoxForm />', () => {
  const meals = [
    { id: 1, name: 'test meal', price: 4000 },
    { id: 2, name: 'test meal', price: 4000 },
  ];
  const props = {
    meals,
    defaultMenuName: 'Menu for Today',
    defaultMenuMealIds: [1, 2],
    handleSubmit: jest.fn(),
    isSubmitting: false,
    error: null,
    action: 'set',
    nextUrl: 'http://testurl/next',
    previousUrl: 'http://testurl.com',
    onPageChange: jest.fn(),
  };
  it('should call handleChange when meal name changes', () => {
    const spy = jest.spyOn(MenuCheckBoxForm.prototype, 'handleChange');
    const wrapper = shallow(<MenuCheckBoxForm {...props} />);
    wrapper.find('input').simulate('change', {
      target: { value: 'Menu for Today' },
    });
    expect(spy).toHaveBeenCalled();
    expect(wrapper.state().name).toEqual('Menu for Today');
  });
  it('should call handleCheck when meal is checked', () => {
    const spy = jest.spyOn(MenuCheckBoxForm.prototype, 'handleCheck');
    const wrapper = shallow(<MenuCheckBoxForm {...props} />);
    wrapper.instance().handleCheck({
      target: {
        checked: true,
        name: 'meal-1',
      },
    });
    expect(spy).toHaveBeenCalled();
    wrapper.instance().handleCheck({
      target: {
        checked: false,
        name: 'meal-1',
      },
    });
    expect(wrapper).toMatchSnapshot();
  });
  it('should call handleSubmit when meal submit button is clicked', () => {
    const submitSpy = jest.spyOn(MenuCheckBoxForm.prototype, 'handleSubmit');
    const handleSubmit = jest.fn();
    const wrapper = shallow(<MenuCheckBoxForm {...{ ...props, handleSubmit }} />);
    wrapper.find('form').simulate('submit', {
      preventDefault: jest.fn(),
    });
    expect(submitSpy).toHaveBeenCalled();
    expect(handleSubmit).toHaveBeenCalled();
  });
  it('should match snapshot', () => {
    let tree = shallow(<MenuCheckBoxForm {...props} />);
    expect(tree).toMatchSnapshot();
    tree = shallow(<MenuCheckBoxForm {...{ ...props, isSubmitting: true }} />);
    expect(tree).toMatchSnapshot();
    tree = shallow(<MenuCheckBoxForm {...{
      ...props,
      error: 'Error submitting form',
      }}
    />);
    expect(tree).toMatchSnapshot();
  });
});
