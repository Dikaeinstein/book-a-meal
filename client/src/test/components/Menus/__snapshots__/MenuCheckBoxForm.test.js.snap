// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuCheckBoxForm /> should call handleCheck when meal is checked 1`] = `
<form
  className="form card"
  onSubmit={[Function]}
>
  <h6
    className="text-center"
    style={
      Object {
        "margin": ".5rem 0 1rem 0",
      }
    }
  >
    Select meals to add to menu
  </h6>
  <div>
    <Label
      className="label label-block"
      htmlFor="name"
    >
      Menu Name:
    </Label>
    <input
      name="name"
      onChange={[Function]}
      placeholder="Enter menu name"
      value="Menu for Today"
    />
  </div>
  <div>
    <h5
      style={
        Object {
          "margin": ".5rem 0",
        }
      }
    >
      Meals
    </h5>
    <Connect(Component)
      checkedMeals={
        Object {
          "meal-1": true,
          "meal-2": true,
        }
      }
      handleCheck={[Function]}
      meals={
        Array [
          Object {
            "id": 1,
            "name": "test meal",
            "price": 4000,
          },
          Object {
            "id": 2,
            "name": "test meal",
            "price": 4000,
          },
        ]
      }
    />
    <Paginate
      nextUrl="http://testurl/next"
      onPageChange={[MockFunction]}
      previousUrl="http://testurl.com"
      style={
        Object {
          "marginTop": "0",
        }
      }
    />
  </div>
  <button
    className="btn btn-default"
    style={
      Object {
        "margin": "0 auto 1rem auto",
      }
    }
    type="submit"
    value="Set Menu"
  >
    set
     Menu
  </button>
</form>
`;

exports[`<MenuCheckBoxForm /> should match snapshot 1`] = `
<form
  className="form card"
  onSubmit={[Function]}
>
  <h6
    className="text-center"
    style={
      Object {
        "margin": ".5rem 0 1rem 0",
      }
    }
  >
    Select meals to add to menu
  </h6>
  <div>
    <Label
      className="label label-block"
      htmlFor="name"
    >
      Menu Name:
    </Label>
    <input
      name="name"
      onChange={[Function]}
      placeholder="Enter menu name"
      value="Menu for Today"
    />
  </div>
  <div>
    <h5
      style={
        Object {
          "margin": ".5rem 0",
        }
      }
    >
      Meals
    </h5>
    <Connect(Component)
      checkedMeals={
        Object {
          "meal-1": true,
          "meal-2": true,
        }
      }
      handleCheck={[Function]}
      meals={
        Array [
          Object {
            "id": 1,
            "name": "test meal",
            "price": 4000,
          },
          Object {
            "id": 2,
            "name": "test meal",
            "price": 4000,
          },
        ]
      }
    />
    <Paginate
      nextUrl="http://testurl/next"
      onPageChange={[MockFunction]}
      previousUrl="http://testurl.com"
      style={
        Object {
          "marginTop": "0",
        }
      }
    />
  </div>
  <button
    className="btn btn-default"
    style={
      Object {
        "margin": "0 auto 1rem auto",
      }
    }
    type="submit"
    value="Set Menu"
  >
    set
     Menu
  </button>
</form>
`;

exports[`<MenuCheckBoxForm /> should match snapshot 2`] = `
<form
  className="form card"
  onSubmit={[Function]}
>
  <h6
    className="text-center"
    style={
      Object {
        "margin": ".5rem 0 1rem 0",
      }
    }
  >
    Select meals to add to menu
  </h6>
  <div>
    <Label
      className="label label-block"
      htmlFor="name"
    >
      Menu Name:
    </Label>
    <input
      name="name"
      onChange={[Function]}
      placeholder="Enter menu name"
      value="Menu for Today"
    />
  </div>
  <div>
    <h5
      style={
        Object {
          "margin": ".5rem 0",
        }
      }
    >
      Meals
    </h5>
    <Connect(Component)
      checkedMeals={
        Object {
          "meal-1": true,
          "meal-2": true,
        }
      }
      handleCheck={[Function]}
      meals={
        Array [
          Object {
            "id": 1,
            "name": "test meal",
            "price": 4000,
          },
          Object {
            "id": 2,
            "name": "test meal",
            "price": 4000,
          },
        ]
      }
    />
    <Paginate
      nextUrl="http://testurl/next"
      onPageChange={[MockFunction]}
      previousUrl="http://testurl.com"
      style={
        Object {
          "marginTop": "0",
        }
      }
    />
  </div>
  <button
    className="btn btn-default"
    style={
      Object {
        "margin": "0 auto 1rem auto",
      }
    }
    type="submit"
    value="Set Menu"
  >
    set
     Menu
  </button>
  <Loading
    text="signing in"
  >
    <ProgressBar />
  </Loading>
</form>
`;

exports[`<MenuCheckBoxForm /> should match snapshot 3`] = `
<form
  className="form card"
  onSubmit={[Function]}
>
  <h6
    className="text-center"
    style={
      Object {
        "margin": ".5rem 0 1rem 0",
      }
    }
  >
    Select meals to add to menu
  </h6>
  <div>
    <Label
      className="label label-block"
      htmlFor="name"
    >
      Menu Name:
    </Label>
    <input
      name="name"
      onChange={[Function]}
      placeholder="Enter menu name"
      value="Menu for Today"
    />
  </div>
  <div>
    <h5
      style={
        Object {
          "margin": ".5rem 0",
        }
      }
    >
      Meals
    </h5>
    <Connect(Component)
      checkedMeals={
        Object {
          "meal-1": true,
          "meal-2": true,
        }
      }
      handleCheck={[Function]}
      meals={
        Array [
          Object {
            "id": 1,
            "name": "test meal",
            "price": 4000,
          },
          Object {
            "id": 2,
            "name": "test meal",
            "price": 4000,
          },
        ]
      }
    />
    <Paginate
      nextUrl="http://testurl/next"
      onPageChange={[MockFunction]}
      previousUrl="http://testurl.com"
      style={
        Object {
          "marginTop": "0",
        }
      }
    />
  </div>
  <button
    className="btn btn-default"
    style={
      Object {
        "margin": "0 auto 1rem auto",
      }
    }
    type="submit"
    value="Set Menu"
  >
    set
     Menu
  </button>
  <div
    className="error"
  >
    Error submitting form
  </div>
</form>
`;
